### ISSUE TITLE
JsonSchema as filter

### ISSUE THREAD

**Issue by gregsdennis** [gregsdennis/Manatee.Json#208]
Many people have requested via multiple channels how JSON Schema could be used as a filter on an instance that returns a new instance with only those portions of the original that match the schema.  I like this idea, but it has been repeatedly rejected as it is decidedly out of scope for the spec.

I'd like to try to implement it.

**Comment by gregsdennis** [comment]
Okay.  So there are some complications that I found just trying to set up a test for how this could work.

Given the following schema

```json
{
  "$schema" : "https://json-schema.org/draft/2019-09/schema",
  "type" : "object",
  "properties" : {
      "integer" : {
          "type" : "integer"
        },
      "numberArray" : {
          "type" : "array",
          "items" : {
              "type" : "number"
            }
        }
    },
  "additionalProperties" : false
}
```
and instance

```json
{
  "integer" : 5,
  "numberArray" : [
      4.5,
      false,
      8.6
    ]
}
```

The results would indicate the following errors:

- `#/numberArray` failed because some items were not numbers
- `#/numberArray/2` failed because `false` is not a number

So what is filtered out here?

1. If we filter top-down, then we remove `numberArray`.  Then the top-level passes, and we return `{"integer": 5}`.
2. If we filter bottom-up, then we remove `false`.  But then `numberArray` passes, so we return `{"integer": 5, "numberArray": [4.5, 8.6]}`.

*BUUUUT...* now suppose we add a `"required": ["numberArray"]` to the schema.

1. If we filter top-down, then we remove `numberArray`.  The the top-level fails, so we have to return... ???  We can't return `null` because that's a valid JSON value that doesn't meet the schema (wrong type).  So the only way to get around this is to return an object that includes the filtered value (if it was possible) along with a boolean indicating whether the filter operation was a success.
2. If we filter bottom-up, then we remove `false`.  But then `numberArray` passes, so we return `{"integer": 5, "numberArray": [4.5, 8.6]}`.  (This is the same.)

***NOTE** Don't take this as implying that bottom-up is the way to go.  If we add a `"minItems": 3` requirement to `numberArray`, then `numberArray` would fail, which would result in the top-level still failing.*

----

***Full disclosure**  This was a spike to see if I could get something working in an effort to move toward standardization of the concept.*

In short there are a lot of issues with how to proceed with filtering.  Trying to standardize this would be a nightmare because the only way to address these issues is to dictate that implementations SHOULD support all of these kinds of scenarios through options.  Then implementations get to decide what they want to implement...

So... no.

**Comment by gregsdennis** [comment]
@handrews this might be useful the next time someone asks for filtering as a schema feature.

**Comment by handrews** [comment]
@gregsdennis yes, thanks!

I definitely see this as out-of-scope for the spec itself, but it could be a separate thing that schema validators can opt to implement (like filling in defaults, which is a bit more straightforward and implemented by multiple packages I believe).
